src/sentry/management/commands/serve_normalize.py

```diff
@@ -0,0 +1,138 @@
+"""
+sentry.management.commands.serve_normalize
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:copyright: (c) 2018 by the Sentry Team, see AUTHORS for more details.
+:license: BSD, see LICENSE for more details.
+"""
+from __future__ import absolute_import, print_function
+
+import SocketServer
+import base64
+import os
+import stat
+import sys
+import time
+import traceback
+import json
+
+from django.core.management.base import BaseCommand, CommandError, make_option
+
+
+class EventNormalizeHandler(SocketServer.BaseRequestHandler):
+    """
+    The request handler class for our server.
+
+    It is instantiated once per connection to the server, and must
+    override the handle() method to implement communication to the
+    client.
+    """
+
+    BUFFER_SIZE = 4096
+
+    def handle(self):
+        chunks = []
+
+        # Receive the data
+        while True:
+            rcvd = self.request.recv(self.BUFFER_SIZE)
+            if rcvd is None:
+                raise ValueError('Received None')
+
+            if not rcvd:
+                break
+            chunks.append(rcvd)
+
+        self.data = ''.join(chunks)
+        sys.stdout.write('> Data received, length: {}\n'.format(len(self.data)))
+
+        response = self.handle_data()
+        self.request.sendall(response)
+        self.request.close()
+
+    def encode(self, data):
+        # Normalized data should be serializable
+        return json.dumps(data)
+
+    def decode(self, message):
+        meta, data_encoded = json.loads(message)
+        data = base64.b64decode(data_encoded)
+        return data, meta
+
+    # Here's where the normalization itself happens
+    def process_event(self, data, meta):
+        from sentry.event_manager import EventManager
+
+        ev = EventManager(
+            data,
+            client_ip=meta.get('REMOTE_ADDR'),
+            user_agent=meta.get('HTTP_USER_AGENT'),
+            auth=None,
+            key=None,
+            content_encoding=meta.get('HTTP_CONTENT_ENCODING')
+        )
+        ev.normalize()
+        return dict(ev.get_data())
+
+    def handle_data(self):
+        result, error = None, None
+        start, end = None, None
+        try:
+            data, meta = self.decode(self.data)
+            start = time.time()
+            result = self.process_event(data, meta)
+            end = time.time()
+        except Exception as e:
+            error = e.message + ' ' + traceback.format_exc()
```

Comment:
Can't this crash with encoding errors as well, due to unicode coercion?

Reply:
I guess you're right, I plugged some `force_str`, just in case.